### ISSUE TITLE
Mixed pagination seems to be not possible in AC3

### ISSUE THREAD

**Issue by Dattaya** [apollographql/apollo-client#7644]
Previously in AC2, if you wanted a 'load more' button in mobile view and regular pagination for larger screens, or the other way around (which is what reddit is doing, next button on smartphones, probably to reduce memory consumption), or both at the same time similar to this one:
![pagination](https://user-images.githubusercontent.com/387256/106652395-11a20100-659e-11eb-85d1-68203503deaf.gif)
normal pagination would just work by receiving variables from query params and load more button was a separate entity with `updateQuery` and did not interfere with that process.
Now, when I add `keyArgs` that ignore offset and limit, regular pages stop working and data remains stale when I navigate with `history.push`.
The only solution that I have in mind that kind of works is to add another variable to the query to be able to distinguish between the two modes, and in `read` depending on the value either return everything or `slice` with offset and limit. And then add another `loadMore` for regular pagination that changes that variable and also calls `fetchMore`. But it looks horrendous, requires another variable that the server does not need, and with this approach `previousData` is empty.

**Comment by benjamn** [comment]
@Dattaya Can you share your type/field policy for this field, including `keyArgs`, `read`, and `merge`?

**Comment by Dattaya** [comment]
Definitely, though there's not much happening there, merge is taken from the docs and read's just as I described:
```js
   test: { // query field
    keyArgs: ['filter', 'sort'],
    merge(existing, incoming, { args }) {
      const items = existing?.items ? existing.items.slice(0) : []
      // Insert the incoming elements in the right places, according to args.
      const end = args.offset + Math.min(args.limit, incoming.items.length)
      for (let i = args.offset; i < end; ++i) {
        items[i] = incoming.items[i - args.offset]
      }
      return { ...incoming, items }
    },
    read(val, { args }) {
      if (args.isPages) {
        return { ...val, items: val?.items?.slice(args.offset, args.offset + args.limit) }
      }
      return val
    },
  },
```

**Comment by dovca** [comment]
I managed to get my mixed pagination working by using a dummy query variable called `listOffset` which tells the `read()` function what to actually return from the cache. "Fetch more" logic works just like the regular pagination - you pass the real offset and limit of new data together with `listOffset`, which would have the value of the original `offset`. Important to note that this doesn't use the `fetchMore()` function.

These are my `read()` and `merge()`:

```js
read(existing, {args: {offset = 0, limit = 0}, variables: {listOffset = offset} = {}}) {
  const start = listOffset;
  const end = limit ? offset + limit : undefined;

  if (!existing || start >= existing.length || end - existing.length === limit) {
    return undefined;
  }

  const page = existing.slice(start, end);

  // Return undefined if the array has holes
  for (let i = 0; i < page.length; i++) {
    if (!page[i]) {
      return undefined;
    }
  }

  return page.length ? page : undefined;
},

merge(existing, incoming, {args: {offset = 0}}) {
  const merged = existing ? existing.slice(0) : [];
  for (let i = 0; i < incoming.length; ++i) {
    merged[offset + i] = incoming[i];
  }
  return merged;
},
```

The `end - existing.length === limit` check is necessary to differentiate between fetching a whole new page and reading the last page, which could have less than `limit` elements. This would break though if you had a misaligned offset, which any sane person wouldn't.

**Comment by bignimbus** [comment]
Hi all 👋🏻  I'm cleaning house a bit and saw that discussion on this issue tapered off a while ago. So I'll close this issue but do feel free to post on our [community forum](https://community.apollographql.com/) with more questions! 🙏🏻
